cmd/shell.go

```diff
@@ -305,9 +314,107 @@ func ReadConfigFile(filename string, out interface{}) error {
 	}
 	defer file.Close()
 
-	decoder := json.NewDecoder(file)
-	decoder.DisallowUnknownFields()
-	return decoder.Decode(out)
+	return decodeJSONStrict(file, out)
+}
+
+// ReadAndValidateConfigFile takes a file path as an argument and attempts to
+// unmarshal the content of the file into a struct containing a configuration of
+// a boulder component. Any config keys in the JSON file which do not correspond
+// to expected keys in the config struct will result in errors. It also
+// validates the config using the struct tags defined in the config struct.
+func ReadAndValidateConfigFile(name, filename string) error {
+	file, err := os.Open(filename)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	if name == "boulder-observer" {
+		// Only the boulder-observer uses YAML config files.
+		return ValidateYAMLConfigByName(name, file)
+	}
+	return ValidateJSONConfigByName(name, file)
+}
+
+// ValidateJSONConfigByName takes a config name and an io.Reader containing a JSON
+// representation of a config and validates the config using the struct tags
+// defined in the config struct.
+func ValidateJSONConfigByName(name string, in io.Reader) error {
+	c := LookupConfig(name)
+	if c == nil {
+		return fmt.Errorf("no config found for %q", name)
+	}
+
+	// Initialize the validator and load any custom tags.
+	validate := validator.New()
+	if c.Validators != nil {
+		for tag, v := range c.Validators {
+			validate.RegisterValidation(tag, v)
+		}
+	}
+
+	err := decodeJSONStrict(in, c.Config)
```

Comment:
This was surprising to me! Looking at the call sites, like

```
cmd.RegisterConfig(cmdName, &cmd.ConfigValidator{Config: &Config{}})
```

I expected that `&Config{}` was just used to provide a type for reflection (as gorp does during tablemap initialization). But the pointer itself actually gets stored in the global registry, and is written to when validating. That has some surprising consequences - for instance if two goroutines are concurrently validating, there will be a data race as they attempt to write to the same thing. Or if someone passes a pointer to a struct that they retain locally, they could incidentally read the contents after a validation.

In practice these things don't happen because each binary only ever calls ReadAndValidateConfigFile once, but it's nice to avoid these tricky problems if we can.

One approach would be to use reflection and store the type value. But if you follow the suggestion above that we rely on `ReadConfigFile` to do validation, that tidily solves the problem. ReadConfigFile always gets a pointer to a config struct, and is expected to write to that config struct, which is exactly what we want to do in this function.

Reply:
Oh nice, this is a good catch! I went ahead and inverted your solution a bit. Rather than eat the Reflect cost at init() time (when storing each config), I construct a new `*ConfigValidator` at lookup and use Reflect to copy the inner `*Config`.

```go
func lookupConfig(name string) *ConfigValidator {
	registry.Lock()
	defer registry.Unlock()
	if registry.configs[name] == nil {
		return nil
	}
	configCopy := reflect.New(reflect.ValueOf(
		registry.configs[name].Config).Elem().Type())
	return &ConfigValidator{
		Config:     configCopy.Interface(),
		Validators: registry.configs[name].Validators,
	}
}
```

Reply:
I actually have to fix this because SRE plans to use the library itself to validate config files.